src/BleOtaUploader.cpp

```diff
@@ -1,288 +1,499 @@
+    if (withChecksum)
     {
-        terminateUpload();
-        handleError(INCORRECT_FIRMWARE_SIZE);
+        if (not _checksum.isSupported())
+        {
+            handleError(BleOtaStatus::ChecksumNotSupported);
+            return;
+        }
+        _checksum.begin();
+    }
+    _checksum.setEnable(withChecksum);
+
+    if (_signature.isEnabled())
+    {
+        _signature.begin();
+    }
+
+    const auto status = _storage.open(req.firmwareSize);
+    if (status != BleOtaStatus::Ok)
+    {
+        handleError(status);
         return;
     }
 
-    uploading = true;
-    crc.restart();
+    _state = State::Upload;
 
-    #ifndef BLE_OTA_NO_BUFFER
-    buffer.clear();
-    uint32_t bufferSize = withBuffer ? BLE_OTA_BUFFER_SIZE : 0;
-    #else
-    uint32_t bufferSize = 0;
-    #endif
-    BeginResponse resp{OK, BLE_OTA_ATTRIBUTE_SIZE, bufferSize};
-    send((const uint8_t*)(&resp), sizeof(BeginResponse));
+    const uint32_t packageSize = min(req.packageSize, BLE_OTA_PACKAGE_SIZE);
+    const uint32_t bufferSize = _buffer.begin(req.bufferSize, packageSize);
+    BLE_OTA_LOG(TAG, "Send BeginResp: size: (pkg: %lu, buff: %lu)", packageSize, bufferSize);
+    sendMessage(BleOtaBeginResp{packageSize, bufferSize});
 
-    ArduinoBleOTA.uploadCallbacks->onBegin(firmwareLength);
+    _uploadCallbacks->handleUploadBegin();
 }
 
-void BleOtaUploader::handlePackage(const uint8_t* data, size_t length)
+void BleOtaUploader::handlePackageReq(const BleOtaPackageReq& req)
 {
-    if (not uploading)
+    BLE_OTA_LOG(TAG, "Handle PackageReq");
+
+    if (_state != State::Upload)
+    {
+        handleError(_state == State::Terminate ? _terminateCode : BleOtaStatus::UploadStopped);
         return;
+    }
 
-    #ifndef BLE_OTA_NO_BUFFER
-    const bool sendResponse = not withBuffer or buffer.size() + length > BLE_OTA_BUFFER_SIZE;
-    #else
-    const bool sendResponse = true;
-    #endif
+#ifndef BLE_OTA_NO_BUFFER
+    BleOtaStatus status;
+    if (_buffer.isEnabled())
+    {
+        status = flushBuffer();
+        if (status == BleOtaStatus::Ok)
+            status = _buffer.push(req.data, req.size);
+    }
+    else
+    {
+        status = pushFirmware(req.data, req.size);
+    }
+#else
+    const auto status = pushFirmware(req.data, req.size);
+#endif
 
-    crc.add(data, length);
-    if (crc.count() > firmwareLength)
+    if (status != BleOtaStatus::Ok)
     {
-        terminateUpload();
-        if (sendResponse) handleError(INCORRECT_FIRMWARE_SIZE);
+        terminateUpload(status);
+        handleError(status);
         return;
     }
 
-    #ifndef BLE_OTA_NO_BUFFER
-    if (sendResponse)
+    sendMessage(BleOtaPackageResp{});
+
+    _uploadCallbacks->handleUploadProgress(_storage.calcProgress());
+}
+
+void BleOtaUploader::handlePackageInd(const BleOtaPackageInd& ind)
+{
+    BLE_OTA_LOG(TAG, "Handle PackageInd");
+
+    if (_state != State::Upload)
+        return;
+
+#ifndef BLE_OTA_NO_BUFFER
```

Comment:
I would use ifdef to avoid double negation

Reply:
It's mot easy, even impossible to use `#ifdef BLE_OTA_NO_BUFFER` in whole file

Reply:
All of that `#ifdef BLE_OTA_NO_..` means cat this code if not used and leave in opposite #ifdef BLE_OTA_(STATIC/DYNAMIC)_..`